---
title: "Plots of ZFTA-RELA ST-EPN patients"
author: "Sara Danielli"
output:
  html_document:
    toc: yes
    df_print: paged
---

```{r, setup, include = FALSE}
library(knitr)
opts_chunk$set(
  echo = TRUE, cache = TRUE, warning = FALSE, comment = FALSE)
```

```{r preparation environment, message=FALSE, warning=FALSE, results='hide'}
# Load packages -----------------------------------
library(tidyverse)
library(ggpubr)
library(dplyr)
library(Seurat)
library(ggplot2)
library(paletteer)
library(readxl)
library(writexl)
library(qs)
library(ggrastr)
library(cowplot)
library(openxlsx)
library(patchwork)

# Organize environment  -----------------------------------
base_dir <- "/Users/sdaniell/Dropbox (Partners HealthCare)/Sara Danielli/Project/Ependymoma/11 - Plots scRNAseq"

data_dir <- file.path(base_dir, "data")
analysis_dir <- file.path(base_dir, 'analysis')
plot_dir <- file.path(analysis_dir, "plots/ZFTA-RELA")
if (!dir.exists(plot_dir)){dir.create(plot_dir, recursive = T)}

ref_dir <- file.path('/Users/sdaniell/Dropbox (Partners HealthCare)/Sara Danielli/Project/Developmental datasets/data/Nowakowski_Eze')

resource_dir <- file.path(base_dir, "scripts/resources")
source(file.path(resource_dir, "Plot_style_v2.R"))
source(file.path(resource_dir, "NMF_helper_function.R")) 
source(file.path(resource_dir, 'single_cell_preprocessing_helper_functions_CBJr.R'))


# Define color palettes  -----------------------------------
col_subtype = c('ZFTA-RELA' = "#B44672",'ZFTA-Cluster 1' = "#B47846", 'ZFTA-Cluster 2' = "#46B478",
                'ZFTA-Cluster 3' = "#46B4AF", 'ZFTA-Cluster 4' = "#4682B4",'ST-YAP1' = "#B4AF46")

colors_metaprograms<- c("gray30","#F99E93FF","#9E5E9BFF","#74ADD1FF",'#0F4F8B', "#ACD39EFF","#96410EFF", 'mistyrose1')

names(colors_metaprograms) <- c("Cycling", "Neuroepithelial-like", "Radial-glia-like", 
                          "Embryonic-neuronal-like", "Neuronal-like" ,"Ependymal-like", "MES-like", "Embryonic-like")

colors_metaprograms_aggregate <- c("gray30","#F99E93FF","#9E5E9BFF","#0F4F8F", "#ACD39EFF","#96410EFF")

names(colors_metaprograms_aggregate) <- c("Cycling", "Neuroepithelial-like", "Radial-glia-like", 
                          "Neuronal-like" ,"Ependymal-like", "MES-like")


colors_metaprograms_Xenium <- c("gray30","#F99E93FF","#9E5E9BFF",
                                '#0F4F8B',"#ACD39EFF","#96410EFF",'grey80', 
                                '#FFF087FF',  'turquoise3', 'turquoise2', 'violetred3', 'violetred2', '#000000FF')

names(colors_metaprograms_Xenium) <- c("Cycling", "Neuroepithelial-like", "Radial glia-like", 
                                       "Neuronal-like" ,"Ependymal-like", "Mesenchymal", "Unassigned", 
                                       "T-cell", "Myeloid", "Microglia", "Endothelial", "VLMCs",  "Oligodendrocyte")


colors_metaprograms_SS2<- c("gray30", "#F99E93FF", "#F99E93FF","#9E5E9BFF", '#0F4F8B', '#0F4F8B',"#ACD39EFF","#96410EFF", 'grey80', '#FFF087FF',
'turquoise3', 'violetred3','#000000FF')

names(colors_metaprograms_SS2) <- c("Cycling", "Embryonic-like",  "Neuroepithelial-like", "Radial-glia-like", 
                          "Neuronal-like" , "Embryonic-neuronal-like", "Ependymal-like", "MES-like", "Unclassified", 
                          "T-cells", "Myeloid", "Endothelial",  "Oligodendrocytes")


# load in data -----------------------------------
# load seurat objects
seurat_obj <- qread(file = file.path(base_dir, "data/patient/seurat_obj_ST_combined_malig_annotated.qs"))

# subset to ZFTA-RELA
Idents(seurat_obj) <- 'Subtype'
seurat_obj <- subset(seurat_obj, idents = 'ZFTA-RELA')

    # Check if the file with tumor aggregated programs already exists
    if (file.exists(file.path(base_dir, "data/patient/seurat_obj_ZFTARELA_combined_malig.qs"))) {
      # If the file exists, print a message or perform any other action
      print("The file already exists. Skipping this part of the code.")
      seurat_obj <- qread(file = file.path(base_dir, "data/patient/seurat_obj_ZFTARELA_combined_malig.qs"))

    } else {
      
      # process with RunFullSeurat
    metadata <- seurat_obj@meta.data
    
    cm <- seurat_obj[["RNA"]]$counts
    cm_norm <- as.matrix(log2(cm/10+1))
    cm_mean <- log2(Matrix::rowMeans(cm)+1)
    cm_center <- cm_norm - rowMeans(cm_norm)
    seurat_obj <- RunFullSeurat_v5(cm = cm, metadata = metadata,  doBatch = T, var2batch = 'sample', batchMethod = 'harmony', project = 'EPN')
    
    # reorder metaprograms
    seurat_obj$Metaprogram <- factor(x = seurat_obj$Metaprogram, levels = c("Cycling", "Embryonic-like", "Neuroepithelial-like", "Radial-glia-like", "Embryonic-neuronal-like", "Neuronal-like"  ,"Ependymal-like", "MES-like"))
    
    # create aggregated metaprograms (aggregate neuronal metaprograms together; neuroepithelial+embryonic)
    seurat_obj[['Metaprogram_aggregate']] = seurat_obj$Metaprogram
    seurat_obj@meta.data$Metaprogram_aggregate <- str_replace_all(seurat_obj@meta.data$Metaprogram_aggregate, "Embryonic-like", "Neuroepithelial-like")
    seurat_obj@meta.data$Metaprogram_aggregate <- str_replace_all(seurat_obj@meta.data$Metaprogram_aggregate, "Embryonic-neuronal-like", "Neuronal-like")
    
    # reorder
    seurat_obj$Metaprogram_aggregate <- factor(x = seurat_obj$Metaprogram_aggregate, 
                                    levels = c("Cycling", "Neuroepithelial-like", "Radial-glia-like", "Neuronal-like"  ,"Ependymal-like", "MES-like")) 
    
    # save
    qsave(seurat_obj, file.path(base_dir, "data/patient/seurat_obj_ZFTARELA_combined_malig.qs"))
    rm(cm, cm_norm, cm_mean, cm_center)
    }
    

# subset to snRNAseq-frozen only (I have normal cells only from this object)
Idents(seurat_obj) <- 'sequencing'
seurat_obj_frozen <- subset(seurat_obj, idents = c('snRNA-seq'))

# load malignant and non malignant object from frozen object
seurat_obj_all <- qread(file = file.path(base_dir, "data/patient/seurat_obj_ST_normal_malig_annotated.qs"))
# keep only ZFTARELA
Idents(seurat_obj_all) <- 'Subtype'
seurat_obj_all_ZFTA <- subset(seurat_obj_all, idents = c('ZFTA-RELA'))
# keep only normal
Idents(seurat_obj_all_ZFTA) <- 'malignant'
seurat_obj_normal_ZFTA <- subset(seurat_obj_all_ZFTA, idents = c('Non-malignant'))
colnames(seurat_obj_normal_ZFTA @meta.data)[20] <- 'Metaprogram'

# concatenate malignant and non-malignant
seurat_obj_frozen_ZFTA_all <- merge(seurat_obj_frozen, seurat_obj_normal_ZFTA)
# reorder Metaprograms
seurat_obj_frozen_ZFTA_all$Metaprogram <- factor(x = seurat_obj_frozen_ZFTA_all$Metaprogram, 
                                levels = c("Cycling", "Neuroepithelial-like", "Embryonic-like", "Radial-glia-like", "Embryonic-neuronal-like", "Neuronal-like" ,"Ependymal-like", "MES-like",  "Endothelial", "Myeloid", "T-cells", "Oligodendrocytes" , "Unclassified")) 



# load NMF gene list
markers_NMF <- readRDS(file.path(base_dir, 'data/signatures/nmf_marker_genes_final_annotated.rds'))

# load Xenium metaprograms
ZFTA_RELA_Xenium <- read.csv(file.path("/Users/sdaniell/Dropbox (Partners HealthCare)/Sara Danielli/Project/Ependymoma/12 - Xenium/analysis/all_Xenium_runs/data/metaprogram_frequency.csv"))

  ZFTA_RELA_Xenium2 <- ZFTA_RELA_Xenium %>% 
    dplyr::group_by(SampleID, group) %>%   
    dplyr::summarise(freq = mean(freq)) %>%  
    ungroup() %>%
    complete(nesting(SampleID), group,
             fill = list(freq = 0)) %>%
    dplyr::mutate(freq2 = freq/100)

  colnames(ZFTA_RELA_Xenium2)[2] <- 'Metaprogram'
  
  ZFTA_RELA_Xenium_average <- ZFTA_RELA_Xenium2 %>% 
    group_by(Metaprogram) %>% 
    summarise(freq = mean(freq2))

  # reorder
  ZFTA_RELA_Xenium_average$Metaprogram <- factor(x = ZFTA_RELA_Xenium_average$Metaprogram,
                                  levels = c("Neuroepithelial-like","Radial glia-like",  "Neuronal-like" ,"Ependymal-like",  "Mesenchymal", "Endothelial", "VLMCs" , "Myeloid", "Microglia" , "T-cell" ,"Oligodendrocytes", "Unassigned" )) 

```

```{r score for cycling, message=FALSE, warning=FALSE, results='hide'}
# Cell cycle scoring -----------------------------------
  ## Cell cycle scoring
  s.genes <- cc.genes$s.genes
  g2m.genes <- cc.genes$g2m.genes
  seurat_obj <- CellCycleScoring(seurat_obj, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
  
  ## Define high-cycling vs low-cycling cells
  Cycling <- WhichCells(seurat_obj, expression = S.Score > 0 | G2M.Score > 0)
  nonCycling <- WhichCells(seurat_obj, expression = S.Score <= 0 & G2M.Score <= 0)
  
  # Add cycling info to metadata column called 'Cycling_prop'
  seurat_obj$Cycling_prop <- ifelse(colnames(seurat_obj) %in% Cycling, "Cycling", "Non-cycling")
  
  # scale data
  seurat_obj <- ScaleData(seurat_obj)
```

# tSNE Plot of frozen samples
## harmony-corrected
```{r UMAP corrected frozen, fig.width = 5, fig.height = 3.5, fig.align="center"}
DimPlot(object = seurat_obj, 
        reduction = 'umap.harmony', 
        group.by = "Metaprogram", 
        pt.size = 0.5, 
        shuffle = TRUE, 
        label = TRUE,
        label.size = 0,
        cols = colors_metaprograms
        ) + 
  labs(title = 'ZFTA-RELA', subtitle = 'n = 2,854 cells') +
  theme_tSNE
ggsave(file.path(plot_dir, "1_umap_harmony_metaprogram.pdf"), width=5, height=3.5)
```


## harmony-corrected
```{r UMAP corrected  aggregate, fig.width = 5, fig.height = 3.5, fig.align="center"}
DimPlot(object = seurat_obj, 
        reduction = 'umap.harmony', 
        group.by = "Metaprogram_aggregate", 
        pt.size = 0.5, 
        shuffle = TRUE, 
        label = TRUE,
        label.size = 0,
        cols = colors_metaprograms
        ) + 
  labs(title = 'ZFTA-RELA', subtitle = 'n = 2,854 cells') +
  theme_tSNE
ggsave(file.path(plot_dir, "1_umap_harmony_metaprogram_aggregate.pdf"), width=5, height=3.5)
```


```{r UMAP sample, fig.width = 4.5, fig.height = 3.5, fig.align="center"}
## UMAP By sample
DimPlot(object = seurat_obj, reduction = 'umap.harmony', group.by = "Sample_deID", label = FALSE, pt.size = 0.5, shuffle = TRUE, label.size = 0, cols = col_Sample_deID)+ 
  labs(title = 'ZFTA-RELA', subtitle = 'n = 2,854 cells') +
  theme_tSNE
ggsave(file.path(plot_dir, '1_umap_harmony_sampleID.pdf'), width = 4.5, height = 3.5) 


```


## uncorrected
```{r UMAP corrected frozen, fig.width = 5, fig.height = 3.5, fig.align="center"}
DimPlot(object = seurat_obj, 
        reduction = 'umap.unintegrated', 
        group.by = "Metaprogram", 
        pt.size = 0.5, 
        shuffle = TRUE, 
        label = TRUE,
        label.size = 4,
        cols = colors_metaprograms
        ) + 
  labs(title = 'ZFTA-RELA', subtitle = 'n = 2,854 cells') +
  theme_tSNE
ggsave(file.path(plot_dir, "3_umap_unintegrated_Metaprogram.pdf"), width=5.5, height=4)
```



# Barplot metaprograms split by sample

```{r barplot, fig.width = 8, fig.height = 4, fig.align="center"}
## Bar plot
plot_bar(seurat_obj, seurat_obj$Sample_deID, seurat_obj$Metaprogram, col = colors_metaprograms) 
ggsave(file.path(plot_dir, "4_Barplot_sample_combined.pdf"), width=8, height=4)

```



# Score ST-EPN cells for developmental programs
```{r scoring metaprograms}
# Score ST-EPN cells for neurodevelopmental genes
scores <- c(markers_NMF['Neuroepithelial-like'], markers_NMF['Ependymal-like'], markers_NMF['NPC-like2'])
names(scores) <- c('Neuroepithelial-like', 'Ependymal-like', 'Neuronal-like')
seurat_obj <- AddModuleScore(object = seurat_obj, assay = 'RNA', features = scores, name = names(scores), seed=5)

# rename metadata names of scores
# identify number of first column with metadata scores
col_start <- length(colnames(seurat_obj@meta.data)) - length(names(scores)) +1
# identify number of last column with metadata scores
col_end <- length(colnames(seurat_obj@meta.data))
# rename columns with score name
colnames(seurat_obj@meta.data)[col_start:col_end] <- names(scores)
    
```


UMAP plot expression of scores
```{r tSNE frozen ependymal markers, fig.width = 12, fig.height = 4, fig.align="center"}
p <- FeaturePlot(seurat_obj, features = names(scores), reduction = "umap.harmony", order = TRUE, combine = FALSE, pt.size = 1)

for(i in 1:length(p)) {
  p[[i]] <- p[[i]] + NoAxes() + paletteer::scale_colour_paletteer_c("viridis::mako", direction = -1) +
      labs (title = names(scores)[[i]]) 
  }
plot_grid(p[[1]], p[[2]], p[[3]], ncol = 3) 
ggsave(file.path(plot_dir, "5_umap_harmony_malignant_scores.pdf"), width=12, height=4, dpi=300)
```

Plot distribution of RNAscope marker marker genes on Dotplot 
```{r dot plot genes, fig.width = 7, fig.height = 3, fig.align="center"}
Idents(seurat_obj) = 'Metaprogram'

  genes <- c("VIM", "TAGLN", "DCX", "SOX11")

DotPlot(seurat_obj, features = genes,
        group.by = 'Metaprogram_aggregate',
        dot.scale = 6,
        #col.min = 0, 
        #cols = c("white", "red3"),
        scale = T, assay = 'RNA') + 
  paletteer::scale_colour_paletteer_c("viridis::mako", direction = -1) + rotate_x_text(45)
ggsave(file.path(plot_dir, "5_Dotplot_markers_RNAscope.pdf"), width=7, height=3)
```

# Barplot cycling cells

```{r barplot cycling, fig.width = 6, fig.height = 4, fig.align="center"}
## Bar plot frozen
seurat_obj$Metaprogram_aggregate <- factor(x = seurat_obj$Metaprogram_aggregate, levels = c( "Cycling",  "Neuroepithelial-like","Ependymal-like", "Neuronal-like" ,  "MES-like",  "Radial-glia-like" ))
plot_bar(seurat_obj, seurat_obj$Metaprogram_aggregate, seurat_obj$Cycling_prop, col_cycling)
ggsave(file.path(plot_dir, "2_Barplot_frozen_cycling.pdf"), width=5, height=4)
```

# Plot primary vs recurrent
Plot bar plot
```{r barplot primary recurrent, fig.width = 5, fig.height = 4, fig.align="center"}
## Bar plot frozen
Idents(seurat_obj) <- 'Sampling'
seurat_obj_subset <- subset(seurat_obj, idents = c('Primary', 'Recurrence'))
plot_bar(seurat_obj_subset, seurat_obj_subset$Sampling, seurat_obj_subset$Metaprogram_aggregate, col = colors_metaprograms_aggregate) 
ggsave(file.path(plot_dir, "2_Barplot_sample_frozen_sampling.pdf"), width=5, height=4)
```

Plot Violin plot
```{r barplot primary recurrent Vln plot, fig.width = 8, fig.height = 3.5, fig.align="center"}
table <- seurat_obj_subset@meta.data %>% 
  dplyr::group_by(sample, Sampling, Metaprogram_aggregate) %>% 
  dplyr::filter(!is.na(Sampling)) %>% 
  dplyr::summarise(n = n()) %>% 
  dplyr::mutate(freq = n / sum(n)) %>%
    ungroup() %>%
    complete(nesting(sample, Sampling,), Metaprogram_aggregate,
             fill = list(n = 0, freq = 0))


## Violin plot
 gg_violinplot_style(table, fill_var = NULL) + 
    geom_violin(aes(fill = Metaprogram_aggregate)) + 
    geom_point(fill = 'black', aes(group = Sampling),size=2,shape=21, position = position_dodge(0.2)) +
    #geom_line(aes(group = PatientID), position = position_dodge(0.2)) +
    labs (y='Proportion', x='') + 
   scale_fill_manual(values=colors_metaprograms_aggregate) + 
    facet_wrap(. ~ Metaprogram_aggregate, nrow = 1, scales = "free") +
    geom_boxplot(width = 0.07) +
   stat_compare_means(aes(label = sprintf("p = %.2f", after_stat(p))), 
                        method = 't.test', 
                        size = 4, 
                        label.y.npc = 0.98, 
                        label.x.npc = 0.5) + NoLegend()
 
ggsave(file.path(plot_dir, "3_VlnPlot_primary_recurrent.pdf"), width=8, height=3.5)

```




# Hierarchy plot of single cells

All cells combined colored by Metaprogram

```{r cellstate plot colored by metaprogram, fig.width = 6, fig.height = 4, fig.align="center"}
## Cell state plot
# x1 <- "Choroid"
# x2 <- "Neuronal_combined"
# y1 <- "Neuroepithelial"
x1 <- "Ependymal-like"
x2 <- "Neuronal-like"
y1 <- "Neuroepithelial-like"
group.by <- "Metaprogram"
variables_to_retrieve <- c(x1, x2, y1, group.by)

# And store them as a tibble.
scores <- seurat_obj@meta.data[, variables_to_retrieve]
scores[["cell"]] <- rownames(scores)
# Shuffle the cells so that we accomplish a random plotting, not sample by sample.
scores <- tidyr::tibble(scores)



# Compute the scores for the X axis.
x <- pbapply::pbvapply(seq_len(nrow(scores)), function(x) {
  score_1 <- scores[x, x1] + stats::runif(1, min=0, max=0.15)
  score_2 <- scores[x, x2] + stats::runif(1, min=0, max=0.15)
  d <- max(score_1, score_2, na.rm = TRUE)
  output <- ifelse(score_1 > score_2, d, -d)
  return(output)
}, FUN.VALUE = numeric(1))


# Compute the scores for the Y axis.
y <- pbapply::pbvapply(seq_len(nrow(scores)), function(x) {
  score_1 <- scores[x, x1] + stats::runif(1, min=0, max=0.15)
  score_2 <- scores[x, x2] + stats::runif(1, min=0, max=0.15)
  d <- max(score_1, score_2, na.rm = TRUE)
  output <- as.data.frame(scores)[x, y1] - d
  return(output)
}, FUN.VALUE = numeric(1))



names(x) <- scores[["cell"]]
names(y) <- scores[["cell"]]

# Define titles.
#lab_x <- paste0(x2, "  <---->  ", x1)
#lab_y <- 'Neuroepithelial/RG-early'

df <- data.frame("set_x" = x, "set_y" = y, "group.by" = scores[[group.by]])

samples = unique(df$group.by)


for (i in 1:length(samples)) {
  tmp <- df$group.by==samples[i]
  tmp <- gsub("TRUE",samples[i],tmp)
  tmp <- gsub("FALSE","Other",tmp)
  df <- cbind(df,tmp)
}
Metaprogram <- unique(seurat_obj@meta.data$Metaprogram)
colnames(df)[4:10] <- Metaprogram

# Define limits of plots
lim <- max(abs(x))
lim_x <- c(-lim, lim)
lim <- max(abs(y))
lim_y <- c(-lim, lim)

#Define colors 
colors.use=colors_metaprograms

df$group.by <- factor(df$group.by, levels = unique(df$group.by))

ggplot(df,aes(x = set_x, y = set_y, col = group.by)) +
  geom_point(size = 0.5)+
  scale_color_manual(values=colors.use)+
  theme_cellstate_plot + 
  labs(x = 'Neuronal-like <------> Ependymal-like', y = 'Neuroepithelial-like', title = 'ST-EPN cellular hierarchy', subtitle = 'n = 2,601 cells') +
    geom_vline(xintercept = 0, linetype="dotted") + geom_hline(yintercept = 0, linetype="dotted") 
    #scale_x_continuous(limits = c(-0.7, 1)) + 
    #scale_y_continuous(limits = c(-1.2, 1.5)) 
ggsave(file.path(plot_dir, "6_CellState_plot_single_cells_Metaprogram.pdf"), width=6.5, height=4.5)

```

# Pie chart of composition

```{r pie chart plot, fig.width = 10, fig.height = 3, fig.align="center"}
# Visualize piechart of ZFTA-RELA tumors (fresh and frozen combined) -----------------------------------    
    # Extract metadata information
    metaprogram_distribution_all <- seurat_obj_frozen_ZFTA_all@meta.data %>% 
      dplyr::group_by(Metaprogram) %>% 
      dplyr::summarise(n = n()) %>% 
      dplyr::mutate(freq = n / sum(n)) 
    
    
    
    # Pie chart plot cell composition
    p1 <- ggplot(metaprogram_distribution_all, aes(x = "", y = freq, fill = Metaprogram)) +
      geom_col() + 
      scale_fill_manual(values = colors_metaprograms_SS2) + 
      coord_polar(theta = "y") + 
      piechart_theme + NoLegend() + piechart_theme +
      geom_text(aes(label = paste0(round(freq*100), "%")), position = position_stack(vjust = 0.5)) +
      ggtitle('Combined')
    

    p5 <- ggplot(ZFTA_RELA_Xenium_average, aes(x = "", y = freq, fill = Metaprogram)) +
      geom_col() + 
      scale_fill_manual(values = colors_metaprograms_Xenium) + 
      coord_polar(theta = "y") + 
      piechart_theme + 
      NoLegend() + 
      piechart_theme +
      geom_text(aes(label = paste0(round(freq*100), "%")), position = position_stack(vjust = 0.5)) +
      ggtitle('Xenium')
  
  

    plot_grid(p1, p5, ncol = 2) 
    ggsave(file.path(plot_dir, "3_piechart.pdf"), width=5, height=3) 
    
    ggplot(ZFTA_RELA_Xenium_average, aes(x = "", y = freq, fill = Metaprogram)) +
      geom_col() + 
      scale_fill_manual(values = colors_metaprograms_Xenium) + 
      coord_polar(theta = "y") + 
      piechart_theme + 
      piechart_theme +
      geom_text(aes(label = paste0(round(freq*100), "%")), position = position_stack(vjust = 0.5)) +
      ggtitle('Xenium')
     ggsave(file.path(plot_dir, "3_piechart_legend.pdf"), width=5, height=3) 
    
```

#Linear regression
```{r linear regression, fig.width = 5, fig.height = 4, fig.align="center"}
# Extract metadata information from SS2 -----------------------------------  
SS2_metaprogram_freq <- seurat_obj_frozen_ZFTA_all@meta.data %>% 
  dplyr::group_by(Metaprogram) %>% 
  dplyr::summarise(n = n()) %>% 
  dplyr::mutate(freq = n / sum(n)) 

# Load Xenium metaprograms -----------------------------------  
ZFTA_RELA_Xenium <- read.csv(file.path("/Users/sdaniell/Dropbox (Partners HealthCare)/Sara Danielli/Project/Ependymoma/12 - Xenium/analysis/all_Xenium_runs/data/metaprogram_frequency.csv"))

  ZFTA_RELA_Xenium2 <- ZFTA_RELA_Xenium %>% 
    dplyr::group_by(SampleID, group) %>%   
    dplyr::summarise(freq = mean(freq)) %>%  
    ungroup() %>%
    complete(nesting(SampleID), group,
             fill = list(freq = 0)) %>%
    dplyr::mutate(freq2 = freq/100)

  colnames(ZFTA_RELA_Xenium2)[2] <- 'Metaprogram'
  
  ZFTA_RELA_Xenium_average <- ZFTA_RELA_Xenium2 %>% 
    group_by(Metaprogram) %>% 
    summarise(freq = mean(freq2))

  # reorder
  ZFTA_RELA_Xenium_average$Metaprogram <- factor(x = ZFTA_RELA_Xenium_average$Metaprogram,
                                  levels = c("Neuroepithelial-like","Radial glia-like",  "Neuronal-like" ,"Ependymal-like",  "Mesenchymal", "Endothelial", "VLMCs" , "Myeloid", "Microglia" , "T-cell" ,"Oligodendrocytes", "Unassigned" )) 
  
  

# Modify names -----------------------------------    
SS2_metaprogram_freq <- SS2_metaprogram_freq[, c('Metaprogram', 'freq')]

SS2_metaprogram_freq$Metaprogram[SS2_metaprogram_freq$Metaprogram == "Embryonic-like"] <- "Neuroepithelial-like"
SS2_metaprogram_freq$Metaprogram[SS2_metaprogram_freq$Metaprogram == "Embryonic-neuronal-like"] <- "Neuronal-like"
SS2_metaprogram_freq$Metaprogram <- gsub("MES-like", "MES/Hypoxia", SS2_metaprogram_freq$Metaprogram)
SS2_metaprogram_freq$Metaprogram <- gsub("Radial-glia-like", "Radial glia-like", SS2_metaprogram_freq$Metaprogram)
SS2_metaprogram_freq$Metaprogram <- gsub("Unclassified", "Unassigned", SS2_metaprogram_freq$Metaprogram)

# sum values of Embryonic and neuroepithelial; and of Embryonic-neuronal and neuronal
SS2_metaprogram_freq <- SS2_metaprogram_freq %>%
  group_by(Metaprogram) %>%
  summarize(total_freq = sum(freq, na.rm = TRUE))

Xenium_metaprogram_freq <- ZFTA_RELA_Xenium_average

Xenium_metaprogram_freq$Metaprogram <- gsub("Mesenchymal", "MES/Hypoxia", Xenium_metaprogram_freq$Metaprogram)
Xenium_metaprogram_freq$Metaprogram <- gsub("Microglia", "Myeloid", Xenium_metaprogram_freq$Metaprogram)
Xenium_metaprogram_freq$Metaprogram <- gsub("VLMCs", "Endothelial", Xenium_metaprogram_freq$Metaprogram)
Xenium_metaprogram_freq$Metaprogram <- gsub("T-cell", "T-cells", Xenium_metaprogram_freq$Metaprogram)

colnames(Xenium_metaprogram_freq)[1] <- 'Metaprogram_Xenium'
colnames(Xenium_metaprogram_freq)[2] <- 'freq_Xenium'

colnames(SS2_metaprogram_freq)[1] <- 'Metaprogram_SS2'
colnames(SS2_metaprogram_freq)[2] <- 'freq_SS2'

# remove cycling (not present in Xenium)
SS2_metaprogram_freq <- SS2_metaprogram_freq %>%
  filter(Metaprogram_SS2 != "Cycling")%>%
  filter(Metaprogram_SS2 != "Unassigned")

# Use match() to get the indices of matching Cluster_names in df2
matching_indices <- match(SS2_metaprogram_freq$Metaprogram_SS2, Xenium_metaprogram_freq$Metaprogram_Xenium)

SS2_metaprogram_freq$Metaprogram_Xenium <- Xenium_metaprogram_freq$Metaprogram_Xenium[matching_indices]
SS2_metaprogram_freq$freq_Xenium <- Xenium_metaprogram_freq$freq_Xenium[matching_indices]


# calculate linear regression  -----------------------------------    
model <- lm(freq_Xenium~freq_SS2, data=SS2_metaprogram_freq)
summary(model)

# R-squared Calculation
# Original data
original_data <- data.frame(x = x, y = y)
# Interpolated data
interpolated_data <- data.frame(x = interp_data$x, y = interp_data$y)

# Plot  -----------------------------------    

colors_regression <- c("gray30", "#F99E93FF", "#F99E93FF","#9E5E9BFF", '#0F4F8B', '#0F4F8B',"#ACD39EFF","#96410EFF",'violetred3',
                       'turquoise3', '#FFF087FF','#000000FF','grey80')

names(colors_regression) <- c("Cycling", "Neuroepithelial-like",  "Neuroepithelial-like", "Radial glia-like", 
                          "Neuronal-like" , "Neuronal-like", "Ependymal-like", "MES/Hypoxia", "Endothelial",
                          "Myeloid", "T-cells","Oligodendrocytes",  "Unclassified")


ggplot(SS2_metaprogram_freq, aes(x = freq_SS2, y = freq_Xenium, color = Metaprogram_SS2)) +
  geom_point(size = 4) + 
  labs(x = "Proportion Smart-seq2", y = "Proportion Xenium", color = 'Metaprogram') +
  scale_color_manual(values = colors_regression) +
  geom_smooth(method='lm', se=T, color='black') +
  theme_minimal() + theme_vln + 
  labs(subtitle = 'R2 = 0.9014, p-value: 9.118e-05')
ggsave(file.path(plot_dir, "4_correlation.pdf"), width=5, height=3.5) 



```


```{r score, warning=FALSE, results= 'hide'}
## Score cells for EPN (ZFTA::RELA) scores identified in patient cohort 
seurat_obj_gene_list
# select top 100 genes per program
ref.gene.EPN <- lapply(seurat_obj_gene_list,head,100)

# score tumors
seurat_obj <- AddModuleScore(object = seurat_obj, assay = 'RNA', features = ref.gene.EPN, name = names(ref.gene.EPN))
seurat_obj <- ScaleData(seurat_obj)
# rename metadata names of scores
  # identify number of first column with metadata scores
  col_start <- length(colnames(seurat_obj@meta.data)) - length(names(ref.gene.EPN)) +1
  # identify number of last column with metadata scores
  col_end <- length(colnames(seurat_obj@meta.data))
  # rename columns with score name
  colnames(seurat_obj@meta.data)[col_start:col_end] <- names(ref.gene.EPN)
```
