---
title: "DEGs of scRNAseq results of mono- vs. co-culture of EP1NS cells"
author: "Sara Danielli"
output:
  html_document:
    toc: yes
    df_print: paged
  html_notebook:
    toc: yes
    toc_float: yes
---

```{r, setup, include = FALSE}
library(knitr)
opts_chunk$set(
  echo = TRUE, cache = TRUE, warning = FALSE, comment = FALSE)
```

```{r, preparation environment}
# Load packages -----------------------------------
library(dplyr)
library(Seurat)
library(ggplot2)
library(patchwork)
library(SeuratObject)
library(data.table)
library(tidyverse)
library(future)
library(clustree)
library(clusterProfiler)
library(ggpubr)
library(cowplot)
library(ggrastr)
library(enrichplot)
library(msigdbr)

# Organize environment  -----------------------------------
base_dir <- "/Users/sdaniell/Dropbox (Partners HealthCare)/Sara Danielli/Project/Ependymoma/9 - scRNAseq coculture"
analysis_dir <- file.path(base_dir, 'analysis')
resource_dir <- file.path(base_dir, 'scripts/resources')

plot_dir <- file.path(analysis_dir, 'plots/DEG')
if (!dir.exists(plot_dir)){dir.create(plot_dir, recursive = T)}
data_dir <- file.path(analysis_dir, 'data')
if (!dir.exists(data_dir)){dir.create(data_dir, recursive = T)}

source(file.path(resource_dir, 'Plot_style_v2.R'))
source(file.path(resource_dir, 'single_cell_preprocessing_helper_functions.R'))


# Define color palettes  -----------------------------------
col_model <- c( 'grey', 'red')
col_cluster_label_transfer <- c("#F99E93FF")


# load Seurat object -----------------------------------
EP1NS_combined <- readRDS(file.path(data_dir, "Mono_co_culture_merged.rds"))
# reorder model order
EP1NS_combined$sample <- factor(x = EP1NS_combined$sample, 
                                  levels = c("Monoculture", "Coculture"))

```

# Identify differentially expressed genes
```{r genes, fig.width = 5, fig.height = 4, fig.align="center"}
Idents(EP1NS_combined) = 'sample' 
markers <- FindAllMarkers(
  EP1NS_combined,
  logfc.threshold = 0.5,
  min.pct = 0.25,
  test.use = "wilcox"
)
markers
write.csv(markers, file.path(plot_dir, "1_markers_coculture_vs_monoculture.csv"))
```
### Exploring the identified differentially expressed genes

```{r volcano, fig.width = 6, fig.height = 5, fig.align="center"}
# subset for markers of coculture only (it is the same as monoculture but with reversed sign)
markers_coculture <- markers %>% 
  filter(cluster %in% c("Coculture"))


EnhancedVolcano::EnhancedVolcano(markers_coculture,
                                 x = 'avg_log2FC', # fold change statistic to plot
                                 y = 'p_val_adj', # significance values
                                 lab = markers_coculture$gene, # labels for points
                                 pCutoff = 1e-05, # p value cutoff (default)
                                 FCcutoff = 1, # fold change cutoff (default)
                                 title = NULL, # no title
                                 subtitle = NULL, # or subtitle
                                 caption = NULL, # or caption
                                 drawConnectors = TRUE, # add some fun arrows
                                 labSize = 3  # smaller labels
) +
  # change the overall theme
  theme_bw() +
  # move the legend to the bottom
  theme(legend.position = "bottom")
ggsave(file.path(plot_dir, "2_Volcano_plot_coculture.pdf"), width=6, height=5)

```
# Perform overrrepresentation analysis 
Now let's find the pathways enriched in up/down regulated genes using ClusterProfiler.

In this notebook, we'll cover a type of pathway or gene set analysis called over-representation analysis (ORA).
The idea behind ORA is relatively straightforward: given a set of genes, do these genes overlap with a pathway more than we expect by chance?
The simplicity of only requiring an input gene set (sort of, more on that below) can be attractive.

ORA has some limitations, outlined nicely (and more extensively!) in [Khatri _et al._ (2012)]( https://doi.org/10.1371/journal.pcbi.1002375).
One of the main issues with ORA is that typically all genes are treated as equal -- the context of the magnitude of a change we may be measuring is removed and each gene is treated as independent, which can sometimes result in an incorrect estimate of significance.

We will use the [`clusterProfiler` package](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) ([Yu *et al.* 2012](https://doi.org/10.1089/omi.2011.0118.)) to perform ORA.
`clusterProfiler` has many built-in functions that will run a specific type of analysis using a specific source of pathways/gene sets automatically, but for our purposes we're going to keep things as general as possible.
See the [`clusterProfiler` book](https://yulab-smu.github.io/clusterProfiler-book/index.html) for more information about the package's full suite of functionality.


## Over-representation Analysis (ORA) of markers upregulated in coculture

### Marker gene preparation
We're interested in what pathways are over-represented in genes that mark, or are more highly expressed in, co-culture

#### Markers of interest set (top 300)
```{r coculture_genes}
markers <- markers_coculture |>
  # Positive fold changes
  dplyr::filter(avg_log2FC > 0) |>
    # Significant p values
  dplyr::filter(p_val_adj < 0.05) |>
    # Take the "top 100" when ordering by FDR (smallest FDR values)
  dplyr::slice_min(order_by = avg_log2FC, n = 300) |>
  # Extract a vector of the gene symbols
  dplyr::pull(gene)
```


### Gene Ontology ORA

The Gene Ontology (GO) ([Ashburner *et al.* 2000](https://dx.doi.org/10.1038/75556); [The Gene Ontology Consortium. 2021](https://doi.org/10.1093/nar/gkaa1113))is [an ontology](https://en.wikipedia.org/wiki/Ontology_(information_science)) that documents or describes how genes function in the biological domain and is comprised of 3 parts ([GO ontology documentation](http://geneontology.org/docs/ontology-documentation/)):

* **Molecular function (MF):** activities that occur on the molecular level, often can be performed by individual genes
* **Biological process (BP):** programs carried out by multiple molecular entities; similar to what is sometimes thought of as a pathway, but with no notion of _dynamics_
* **Cellular component (CC):** cellular structures and compartments

GO is "loosely hierarchical" ([GO ontology documentation](http://geneontology.org/docs/ontology-documentation/)).
There are parent and child terms, where the "higher up" a term is, the more general it is.
This is somewhat abstract, so let's take a look at a GO term that's likely to be relevant to our Hodgkin's lymphoma data to get an idea of what this is like in practice: [`leukocyte activation`](http://www.informatics.jax.org/vocab/gene_ontology/GO:0045321).

#### Run `enrichGO()`

Now that we have our background set and our genes of interest, we're ready to run ORA using the `enrichGO()` function.

```{r go_ora2}
go_ora_results <- enrichGO(gene = markers,  # Top 300 genes
                           #universe = background_set,  # Background genes
                           keyType = "SYMBOL",  # Our genes are gene symbols
                           OrgDb = org.Hs.eg.db,  # Supports conversion, if needed
                           ont = "BP",  # Biological process
                           pAdjustMethod = "BH",  # FDR
                           pvalueCutoff = 0.001) 
```


The information we're most likely interested in is in the `result` slot.
Let's convert this into a data frame that we can write to file.

```{r go_df2}
go_result_df <- data.frame(go_ora_results@result)
write.csv(go_result_df, file.path(plot_dir, "3_enrichGO_coculture.csv"))

```

Let's take a look at the GO sets with an adjusted p-value less than 0.001.

```{r filter_padj2, live = TRUE}
go_result_df <- go_result_df |>
  dplyr::filter(p.adjust < 0.001)
write.csv(go_result_df, file.path(plot_dir, "4_enrichGO_coculture_pval_001.csv"))

```

Some of these columns have names that are somewhat opaque and are used for visualization, so let's briefly talk about them here!

| Column name | Numerator | Denominator |
|-------------|-----------|-------------|
| `GeneRatio` | Number of genes in the top 100 marker genes and the specific gene set (also in `Count` column) | Number of genes in the top 100 marker genes and _any_ gene set |
| `BgRatio`   | Number of genes in the background set and the specific gene set | Number of genes in the background set and _any_ gene set |

#### Visualizing results

We can use a dot plot to visualize our significant enrichment results.

```{r dotplot2, live = TRUE, fig.width = 6, fig.height = 10, fig.align="center"}
enrichplot::dotplot(go_ora_results,  showCategory = 20)
ggsave(file.path(plot_dir, "5_Dotplot_coculture.pdf"), width=6, height=10)
```

## Over-representation Analysis (ORA) of markers downregulated in coculture

### Marker gene preparation
We're interested in what pathways are over-represented in genes that mark, or are more highly expressed in, co-culture

#### Markers of interest set (top 300)
```{r FBS_genes}
markers_down <- markers_coculture |>
  # Positive fold changes
  dplyr::filter(avg_log2FC < 0) |>
    # Significant p values
  dplyr::filter(p_val_adj < 0.05) |>
    # Take the "top 100" when ordering by FDR (smallest FDR values)
  dplyr::slice_min(order_by = avg_log2FC, n = 300) |>
  # Extract a vector of the gene symbols
  dplyr::pull(gene)
```


### Gene Ontology ORA

The Gene Ontology (GO) ([Ashburner *et al.* 2000](https://dx.doi.org/10.1038/75556); [The Gene Ontology Consortium. 2021](https://doi.org/10.1093/nar/gkaa1113))is [an ontology](https://en.wikipedia.org/wiki/Ontology_(information_science)) that documents or describes how genes function in the biological domain and is comprised of 3 parts ([GO ontology documentation](http://geneontology.org/docs/ontology-documentation/)):

* **Molecular function (MF):** activities that occur on the molecular level, often can be performed by individual genes
* **Biological process (BP):** programs carried out by multiple molecular entities; similar to what is sometimes thought of as a pathway, but with no notion of _dynamics_
* **Cellular component (CC):** cellular structures and compartments

GO is "loosely hierarchical" ([GO ontology documentation](http://geneontology.org/docs/ontology-documentation/)).
There are parent and child terms, where the "higher up" a term is, the more general it is.
This is somewhat abstract, so let's take a look at a GO term that's likely to be relevant to our Hodgkin's lymphoma data to get an idea of what this is like in practice: [`leukocyte activation`](http://www.informatics.jax.org/vocab/gene_ontology/GO:0045321).

#### Run `enrichGO()`

Now that we have our background set and our genes of interest, we're ready to run ORA using the `enrichGO()` function.

```{r go_ora}
go_ora_results_down <- enrichGO(gene = markers_down,  # Top 300 genes
                           #universe = background_set,  # Background genes
                           keyType = "SYMBOL",  # Our genes are gene symbols
                           OrgDb = org.Hs.eg.db,  # Supports conversion, if needed
                           ont = "BP",  # Biological process
                           pAdjustMethod = "BH",  # FDR
                           pvalueCutoff = 0.001) 
```


The information we're most likely interested in is in the `result` slot.
Let's convert this into a data frame that we can write to file.

```{r go_df}
go_result_df_down <- data.frame(go_ora_results_down@result)
write.csv(go_result_df_down, file.path(plot_dir, "6_enrichGO_monoculture.csv"))

```

Let's take a look at the GO sets with an adjusted p-value less than 0.001.

```{r filter_padj, live = TRUE}
go_result_df_down <- go_result_df_down |>
  dplyr::filter(p.adjust < 0.001)
write.csv(go_result_df_down, file.path(plot_dir, "7_enrichGO_monoculture_pval_001.csv"))

```

Some of these columns have names that are somewhat opaque and are used for visualization, so let's briefly talk about them here!

| Column name | Numerator | Denominator |
|-------------|-----------|-------------|
| `GeneRatio` | Number of genes in the top 100 marker genes and the specific gene set (also in `Count` column) | Number of genes in the top 100 marker genes and _any_ gene set |
| `BgRatio`   | Number of genes in the background set and the specific gene set | Number of genes in the background set and _any_ gene set |

#### Visualizing results

We can use a dot plot to visualize our significant enrichment results.

```{r dotplot, live = TRUE, fig.width = 6, fig.height = 10, fig.align="center"}
enrichplot::dotplot(go_ora_results_down,  showCategory = 20)
ggsave(file.path(plot_dir, "8_Dotplot_monoculture.pdf"), width=6, height=10)
```



## Gene set enrichment analysis (GSEA)
Here, we'll analyze the marker genes enriched in FBS, just as we did in the previous part
Unlike ORA, GSEA allows us to use the full list of genes we could reliably measure, rather than picking some cutoff ourselves.

GSEA is a functional class scoring (FCS) approach to pathway analysis that was first introduced in [Subramanian _et al._ (2005)](https://doi.org/10.1073/pnas.0506580102).
The rationale behind FCS approaches is that small changes in individual genes that participate in the same biological process or pathway can be significant and of biological interest.
FCS methods are better suited for identifying these pathways that show coordinated changes than ORA.
In ORA, we pick a cutoff that _typically_ only captures genes with large individual changes.

There are 3 general steps in FCS methods ([Khatri _et al._ 2012]( https://doi.org/10.1371/journal.pcbi.1002375)):

1. Calculate a gene-level statistic (here, we'll use the summary log fold changes in our marker genes table)
2. Gene-level statistics are aggregated into a pathway-level statistic
3. Assess the statistical significance of the pathway-level statistic

We'll note here that GSEA was designed to detect small coordinated changes between _conditions_ in (necessarily bulk!) microarray data.
It may be more difficult to suss out small coordinated changes when we assay individual cells, due to technical or biological dropout.

Individual cells may also differently sized ([Blasi *et al.* 2017](https://doi.org/10.1088/1478-3975/aa609a)), be at different stages of the cell cycle ([Satija lab](https://satijalab.org/seurat/archive/v3.1/cell_cycle_vignette.html)), or experiencing different levels of cellular stress ([Luecken and Theis. 2019](https://doi.org/10.15252/msb.20188746)), some of which may be interesting in the context of cancer and therefore should not be corrected for.
These aspects also have the potential to cloud our ability to detect small coordinated changes in other pathways.

There is some literature that suggests that dropout (excessive zeros) observed in single cell data is a function of cell type heterogeneity ([Kim *et al.* 2020](https://doi.org/10.1186/s13059-020-02096-y)).
We're using GSEA to compare clusters of cells, which we (ideally) expect to be different biological states and/or cell types.
If many genes "drop out" between cell types, we may expect our statistics used as input to GSEA to be noisy.

With all this in mind, it may be unsurprising that ORA, which generally captures genes with large individual changes and lower within-group variation (the most "statistically significant" genes), is well-suited to analyzing scRNA-seq data and the advantages of GSEA in a bulk analysis setting may not be fully available here.
Nonetheless, GSEA is used for pathway analysis of scRNA-seq data.

Methods specifically for pathway analysis of scRNA-seq (e.g., [Ma *et al.* 2020](https://doi.org/10.1038/s41467-020-15298-6)) are being developed, so stay tuned!


### Setup

#### Input files

```{r input_files}
# subset for markers of coculture only (it is the same as monoculture but with reversed sign)
input_file <- markers_coculture %>% 
  filter(p_val_adj < 0.05)

input_file 
```


#### Output files

We'll save our table of GSEA results as a TSV.

```{r output_files}
output_file <- file.path(plot_dir,
                         "9_Coculture_gsea_results.tsv")
```

## Gene sets

We will use gene sets from the [Molecular Signatures Database (MSigDB)](https://www.gsea-msigdb.org/gsea/msigdb/index.jsp) from the Broad Institute ([Subramanian, Tamayo *et al.* 2005](https://doi.org/10.1073/pnas.0506580102)).
The [`msigdbr`](https://cran.r-project.org/web/packages/msigdbr/index.html) package contains MSigDB datasets already in the tidy format required by `clusterProfiler` and supports multiple organisms.

Let's take a look at what organisms the package supports.

```{r show_species}
msigdbr_species()
```

MSigDB contains 8 different gene set collections.

    H: hallmark gene sets
    C1: positional gene sets
    C2: curated gene sets
    C3: motif gene sets
    C4: computational gene sets
    C5: GO gene sets
    C6: oncogenic signatures
    C7: immunologic signatures
    C8: cell type signatures derived from scRNAseq


We can retrieve only specific gene sets by specifying e.g. `category = "C7"` to the `msigdbr()` function.
Again, we only want human gene sets here so we specify with that with the `species` argument.

```{r GSEA gene set, live = TRUE}
hs_cell_type_df <- msigdbr(species = "Homo sapiens",
                             category = "C5")
```

## Gene Set Enrichment Analysis

GSEA calculates a pathway-level metric, called an enrichment score (sometimes abbreviated as ES), by ranking genes by a gene-level statistic.
This score reflects whether or not a gene set or pathway is over-represented at the top or bottom of the gene rankings ([Subramanian _et al._ 2005](https://doi.org/10.1073/pnas.0506580102); [Yu](http://yulab-smu.top/clusterProfiler-book/chapter2.html#gene-set-enrichment-analysis))

Specifically, all genes are ranked from most positive to most negative based on their statistic and a running sum is calculated:
Starting with the most highly ranked genes, the running sum increases for each gene in the pathway and decreases for each gene not in the pathway.
The enrichment score for a pathway is the running sum's maximum deviation from zero.
GSEA also assesses statistical significance of the scores for each pathway through permutation testing.
As a result, each input pathway will have a p-value associated with it that is then corrected for multiple hypothesis testing ([Subramanian _et al._ 2005](https://doi.org/10.1073/pnas.0506580102); [Yu](http://yulab-smu.top/clusterProfiler-book/chapter2.html#gene-set-enrichment-analysis)).

The implementation of GSEA we use in here examples requires a gene list ordered by some statistic and input gene sets.
When you use previously computed gene-level statistics with GSEA, it is called GSEA pre-ranked.

### Removing duplicates 

```{r}
head(input_file)
```

We do need to check for duplicate gene symbols.
We can accomplish this with `duplicated()`, which returns a logical vector (e.g., `TRUE` or `FALSE`).
The function `sum()` will count `TRUE` values as 1s and `FALSE` as 0s, so using it with `duplicated()` will count the number of duplicate values.

```{r any_duplicated, live = TRUE}
sum(duplicated(input_file$gene))
```

Luckily this number is 0

Now we're ready to prep our pre-ranked list for GSEA.

### Pre-ranked list

The `GSEA()` function takes a pre-ranked (sorted) named vector of statistics, where the names in the vector are gene identifiers.
This is step 1 -- gene-level statistics.

Here, we're using the summary log fold change, which summarizes of all of the log fold changes from the pairwise comparisons between cluster 1 and other clusters into a single value ([ref](https://rdrr.io/bioc/scran/man/combineMarkers.html)).
As such, this statistic gives us information about the relative magnitude and directionality of each gene's expression in cluster 1 relative to all other clusters.


```{r lfc_vector}
lfc_vector <- input_file |>
  # Extract a vector of `summary.logFC` named by `gene_symbol`
  dplyr::pull(avg_log2FC, name = gene)
lfc_vector <- sort(lfc_vector, decreasing = TRUE)
```

Let's look at the top ranked values.

```{r head_lfc, live = TRUE}
# Look at first entries of the log fold change vector
head(lfc_vector)
```

And the bottom of the list.

```{r tail_lfc, live = TRUE}
# Look at the last entries of the log fold change vector
tail(lfc_vector)
```

## Run GSEA

Now for the analysis!

We can use the `GSEA()` function to perform GSEA with any generic set of gene sets, but there are several functions for using specific, commonly used gene sets (e.g., `gseKEGG()`).

```{r run_gsea}
gsea_results <- GSEA(geneList = lfc_vector,  # ordered ranked gene list
                     minGSSize = 5,  # minimum gene set size
                     maxGSSize = 500,  # maximum gene set set
                     pvalueCutoff = 0.05,
                     pAdjustMethod = "BH",  # correction for multiple hypothesis testing
                     TERM2GENE = dplyr::select(hs_cell_type_df,
                                               gs_name,
                                               gene_symbol))
```
Let's take a look at the GSEA results.

```{r view_gsea2, live = TRUE, eval = FALSE}
View(gsea_results@result |>
       dplyr::arrange(dplyr::desc(NES))
)
```

Normalized enrichment scores (NES) are enrichment scores that are scaled to make gene sets that contain different number of genes comparable.

Let's write these results to file.

```{r write_gsea}
gsea_results@result |> readr::write_tsv(output_file)
```

Let's take a look at the GSEA results.

```{r view_gsea, live = TRUE, eval = FALSE}
View(gsea_results@result |>
       dplyr::arrange(dplyr::desc(NES))
)
```

And visualize the top 30 pathways

```{r pathways up-down-regulated, fig.width = 10, fig.height = 10, fig.align="center"}
## Plot barplot with sign of FC
ggplot(gsea_results, aes(NES, fct_reorder(Description, NES), fill=qvalue), showCategory=20) + 
  geom_col(orientation='y') + 
  paletteer::scale_fill_paletteer_c("ggthemes::Brown")  + ylab(NULL) + theme_classic() + 
  geom_vline(xintercept = 0, color = "black") +
  theme(
    axis.text.x = element_text(size = 12, color = "black"),   # Adjust x-axis label properties
    axis.text.y = element_text(size = 12, color = "black"),    # Adjust y-axis label properties
    legend.text = element_text(size = 12, color = "black"),
    legend.title = element_text(size = 12, color = "black")
  )
ggsave(file.path(plot_dir, "10_Barplot.pdf"), width=10, height=10)

```

## Print session info

```{r session info}
sessionInfo()



