---
title: "scRNAseq results of mono- vs. co-culture of EP1NS cells"
author: "Sara Danielli"
output:
  html_document:
    toc: yes
    df_print: paged
---

```{r, setup, include = FALSE}
library(knitr)
opts_chunk$set(
  echo = TRUE, cache = TRUE, warning = FALSE, comment = FALSE)
```

```{r preparation environment, message=FALSE, warning=FALSE}
# Load packages -----------------------------------
library(dplyr)
library(Seurat)
library(ggplot2)
library(patchwork)
library(SeuratObject)
library(data.table)
library(tidyverse)
library(future)
#library(clustree)
#library(clusterProfiler)
library(ggpubr)
library(cowplot)
library(ggrastr)

# Organize environment  -----------------------------------
base_dir <- "/Users/sdaniell/Dropbox (Partners HealthCare)/Sara Danielli/Project/Ependymoma/9 - scRNAseq coculture"
analysis_dir <- file.path(base_dir, 'analysis')
resource_dir <- file.path(base_dir, 'scripts/resources')

plot_dir <- file.path(analysis_dir, 'plots/mono_vs_coculture')
if (!dir.exists(plot_dir)){dir.create(plot_dir, recursive = T)}
data_dir <- file.path(analysis_dir, 'data')
if (!dir.exists(data_dir)){dir.create(data_dir, recursive = T)}

source(file.path(resource_dir, 'Plot_style_v2.R'))
source(file.path(resource_dir, 'single_cell_preprocessing_helper_functions.R'))


# Define color palettes  -----------------------------------
colors_metaprograms<- c("gray30","#F99E93FF","#9E5E9BFF","#74ADD1FF",'#0F4F8B', "#ACD39EFF","#96410EFF", 'mistyrose1')

names(colors_metaprograms) <- c("Cycling", "Neuroepithelial-like", "Radial-glia-like", 
                          "Embryonic-neuronal-like", "Neuronal-like" ,"Ependymal-like", "MES-like", "Embryonic-like")


col_model <- c('grey80', 'grey20', 'red')
names(col_model) <- c('Patient', 'Monoculture', 'Coculture')

# load Seurat object -----------------------------------
EP1NS_combined <- readRDS(file.path(data_dir, "Mono_co_culture_merged.rds"))
# reorder model order
EP1NS_combined$sample <- factor(x = EP1NS_combined$sample, 
                                  levels = c("Monoculture", "Coculture"))
# reorder label-transfer cluster order
EP1NS_combined$Metaprogram <- factor(x = EP1NS_combined$Metaprogram, levels = c("Cycling", "Embryonic-like", "Neuroepithelial-like", "Radial-glia-like", 
                                                                                "Embryonic-neuronal-like", "Neuronal-like"  ,"Ependymal-like", "MES-like")) 

EP1NS_patient_combined <- readRDS(file.path(data_dir, "Mono_co_culture_patient_merged.rds"))

```

# UMAP plots

Plot UMAP colored by different metadata columns


## By sample
```{r UMAP sample, fig.width = 5, fig.height = 4, fig.align="center"}
DimPlot(object = EP1NS_combined, 
        reduction = 'umap', 
        group.by = "sample", 
        pt.size = 2, 
        shuffle = TRUE, 
        label = TRUE,
        label.size = 0,
        cols = col_model
        ) + 
  labs(title = 'EP1NS', subtitle = 'n = 919 cells') 
ggsave(file.path(plot_dir, "2_umap_sample.pdf"), width=5, height=4)

```  

## By Metaprogram
```{r UMAP metaprogram, fig.width = 6, fig.height = 4, fig.align="center"}
DimPlot(object = EP1NS_combined, 
        reduction = 'umap', 
        group.by = "Metaprogram", 
        pt.size =2, 
        shuffle = TRUE, 
        label = TRUE,
        label.size = 0,
        cols = colors_metaprograms
        ) + 
  labs(title = 'EP1NS', subtitle = 'n = 919 cells') 
ggsave(file.path(plot_dir, "3_umap_metaprograms.pdf"), width=6, height=4)
```  

## By scores
```{r UMAP scores, fig.width = 9, fig.height = 6, fig.align="center"}
## by patient scores
scores <- c("Neuroepithelial-like", "MES-like","Ependymal-like","Radial-glia-like","Cycling", "Embryonic-neuronal-like", "Neuronal-like", "Embryonic-like")
p <- FeaturePlot(EP1NS_combined, features = scores, reduction = "umap", combine = FALSE, pt.size = 1)

for(i in 1:length(p)) {
  p[[i]] <- p[[i]] + NoAxes() + scale_colour_distiller(palette="RdBu")
  }
cowplot::plot_grid(plotlist = p, ncol=3)
ggsave(file.path(plot_dir, "4_umap_scores.pdf"), width=9, height=6)
```   

```{r UMAP scores2, fig.width = 10, fig.height = 3, fig.align="center"}
plot_grid(p[[1]], p[[7]], p[[3]], ncol = 3) 
ggsave(file.path(plot_dir, "4_umap_scores_selected.pdf"), width=10, height=3, dpi=300)
```   

# Violin plots
Plot Violin plots of scores 

```{r Vln plots, fig.width = 10, fig.height = 8, fig.align="center"}
titles <- scores

p <- VlnPlot(EP1NS_combined, features = scores, group.by = 'sample', combine = FALSE, pt.size=0, cols = col_model) 

for(i in 1:length(p)) {
    p[[i]] <- p[[i]] + NoLegend() +
      labs (y='Module score, AU', x='', title = titles[[i]]) + 
      scale_fill_manual(values = col_model) + 
      geom_boxplot(outlier.shape=NA, width=0.1, fill="white") +
      stat_compare_means(aes(label = after_stat(p.format)), 
                         method = 't.test', 
                         size = 5, 
                         label.y.npc = 0.95, 
                         label.x.npc = 0.5, 
                         ) + NoLegend() +
      theme_vln
  }
plot_grid(plotlist = p, ncol=4)
ggsave(file.path(plot_dir, "5_VlnPlot_scores.pdf"), width=10, height=8)
```

Plot Violin plots of selected scores 

```{r Vln plots 2, fig.width = 5, fig.height = 3, fig.align="center"}
plot_grid(p[[1]], p[[7]], p[[3]], ncol = 3) 
ggsave(file.path(plot_dir, "6_VlnPlot_scores_selected.pdf"), width=6, height=4)
```


# Add module score neurexin pathway
```{r Vln plot markers, fig.width = 4, fig.height = 4, fig.align="center"}
neurexin <- list(c("NRXN1","NLGN1","LRRTM4","NRXN3","DAG1","CLSTN1","NLGN3","CLSTN2","NLGN2", "CLSTN3"))
names(neurexin) <- 'Neurexin'
EP1NS_combined <- AddModuleScore(object = EP1NS_combined, assay = 'RNA', features = neurexin, name = names(neurexin), seed=5)

p <- VlnPlot(EP1NS_combined, features = 'Neurexin1', group.by = 'sample', combine = FALSE, pt.size=0, cols = col_model) 
for(i in 1:length(p)) {
    p[[i]] <- p[[i]] + NoLegend() +
      labs (y='Module score', x='') + 
      scale_fill_manual(values = col_model) + 
      geom_boxplot(outlier.shape=NA, width=0.1, fill="white") +
      stat_compare_means(aes(label = after_stat(p.format)), 
                         method = 't.test', 
                         label.x.npc = 0.5, 
                         label.y.npc = 0.92, 
                         label.x = 1.3
                         ) + 
         theme_vln +
      NoLegend()
  }
plot_grid(plotlist = p, ncol=1)
ggsave(file.path(plot_dir, "6_VlnPlot_neurexin.pdf"), width=4, height=4)

```



# Violin plots
Plot Violin plots of markers neurexin

```{r Vln plot markers, fig.width = 12, fig.height = 8, fig.align="center"}
markers <- c("NRXN1","NLGN1","LRRTM4","NRXN3","DAG1","CLSTN1","NLGN3","CLSTN2","NLGN2", "CLSTN3")

p <- VlnPlot(EP1NS_combined, features = markers, group.by = 'sample', combine = FALSE, pt.size=0, cols = col_model) 

for(i in 1:length(p)) {
    p[[i]] <- p[[i]] + NoLegend() +
      labs (y='Gene expression', x='') + 
      scale_fill_manual(values = col_model) + 
      geom_boxplot(outlier.shape=NA, width=0.1, fill="white") +
      stat_compare_means(aes(label = after_stat(p.format)), 
                         method = 't.test', 
                         label.x.npc = 0.5, 
                         label.y.npc = 0.92, 
                         label.x = 1.3
                         ) + 
         theme_vln +
      NoLegend()
  }
plot_grid(plotlist = p, ncol=5)
ggsave(file.path(plot_dir, "7_VlnPlot_markers_neurexin.pdf"), width=12, height=8)
```

Plot Violin plots of markers 

```{r Vln plot markers, fig.width = 12, fig.height = 3, fig.align="center"}
markers <- c('VIM', 'TAGLN2', 'NES', 'DCX', 'TUBB3', 'SOX11', 'CCDC40', 'S100B')

p <- VlnPlot(EP1NS_combined, features = markers, group.by = 'sample', combine = FALSE, pt.size=0, cols = col_model) 

for(i in 1:length(p)) {
    p[[i]] <- p[[i]] + NoLegend() +
      labs (y='Gene expression', x='') + 
      scale_fill_manual(values = col_model) + 
      geom_boxplot(outlier.shape=NA, width=0.1, fill="white") +
      stat_compare_means(aes(label = after_stat(p.format)), 
                         method = 't.test', 
                         label.x.npc = 0.5, 
                         label.y.npc = 0.92, 
                         label.x = 1.3
                         ) + 
         theme_vln +
      NoLegend()
  }
plot_grid(plotlist = p, ncol=8)
ggsave(file.path(plot_dir, "7_VlnPlot_markers.pdf"), width=12, height=3)
```

Plot Violin plots of markers for Varun

```{r Vln plot markers2, fig.width = 5, fig.height = 3, fig.align="center"}
markers <- c('VIM', 'DCX', 'GFAP')

p <- VlnPlot(EP1NS_combined, features = markers, group.by = 'sample', combine = FALSE, pt.size=0, cols = col_model) 

for(i in 1:length(p)) {
    p[[i]] <- p[[i]] + NoLegend() +
      labs (y='Gene expression', x='') + 
      scale_fill_manual(values = col_model) + 
      geom_boxplot(outlier.shape=NA, width=0.1, fill="white") +
      stat_compare_means(aes(label = after_stat(p.format)), 
                         method = 't.test', 
                         label.x.npc = 0.5, 
                         label.y.npc = 0.92, 
                         label.x = 1.3
                         ) + 
         theme_vln +
      NoLegend()
  }
plot_grid(plotlist = p, ncol=3)
ggsave(file.path(plot_dir, "7_VlnPlot_markers2.pdf"), width=5, height=3.5)
```

# Barplot 

Barplot of label transfer
```{r barplot singleR, fig.width = 5, fig.height = 4, fig.align="center"}
## Bar plot
plot_bar(EP1NS_combined, EP1NS_combined$sample, EP1NS_combined$Metaprogram, col = colors_metaprograms) 
ggsave(file.path(plot_dir, "8_Barplot_label_transfer_sample.pdf"), width=5, height=4)
```

Barplot of high vs low cycling cells
```{r barplot cycling cells, fig.width = 4, fig.height = 4, fig.align="center"}
## Bar plot
plot_bar(EP1NS_combined, EP1NS_combined$sample, EP1NS_combined$Cycling_prop, col = c('red', 'black')) 
ggsave(file.path(plot_dir, "8b_Barplot_cycling.pdf"), width=4, height=4)
```


# Cell state plot

Plot distribution of samples on the hierarchy plot

```{r cellstate plot colored by subtype, fig.width = 6, fig.height = 4, fig.align="center"}
x1 = "Neuroepithelial-like"
x2 = "Embryonic-like"
y1 = "Neuronal-like"
y2 = "Ependymal-like"

group.by="sample"
sample=EP1NS_combined
variables_to_retrieve <- c(x1, x2, y1,y2)

# And store them as a tibble.
scores <- sample@meta.data[, variables_to_retrieve]
# Shuffle the cells so that we accomplish a random plotting, not sample by sample.

# Compute Y axis values.
d <- apply(scores, 1, function(x){max(x[c(x1, x2)]) - max(x[c(y1, y2)])})

# Compute X axis values.
x <- vapply(seq_along(d), function(x) {
  if (d[x] > 0) {
    d <- log2(abs(scores[x, x1] - scores[x, x2]) + 1)
    ifelse(scores[x, x1] < scores[x, x2], d, -d)
  } else {
    d <- log2(abs(scores[x, y1] - scores[x, y2]) + 1)
    ifelse(scores[x, y1] < scores[x, y2], d, -d)
  }
}, FUN.VALUE = numeric(1))

names(x) <- rownames(scores)

# Plot.
df_E <- data.frame(row.names = rownames(scores))
df_E[["set_x"]] <- x
df_E[["set_y"]] <- d
df_E[["group.by"]] <- sample@meta.data[, group.by]
df_E[["subtype"]] <- EP1NS_combined$sample


ggplot(df_E,aes(x=set_x,
              y=set_y,
              col=group.by))+
  geom_point(size=2)+
  scale_color_manual(values=col_model)+
   theme_cellstate_plot + 
  labs(x = 'Relative meta-module score', y = 'Relative meta-module score', title = 'EP1NS cellular hierarchy', subtitle = 'n = 919 cells') +
    geom_vline(xintercept = 0, linetype="dotted") + geom_hline(yintercept = 0, linetype="dotted") +
  xlim(-2, 1) + ylim(-1, 3)

ggsave(file.path(plot_dir, "9_CellState_plot_4way_Metaprograms.pdf"), width=5.5, height=4)

```


# Cell state plot with patients

Plot distribution of samples on the hierarchy plot

```{r cellstate plot colored by subtype, fig.width = 6, fig.height = 4, fig.align="center"}
x1 = "Neuroepithelial-like"
x2 = "Embryonic-like"
y1 = "Neuronal-like"
y2 = "Ependymal-like"

group.by="model"
sample=EP1NS_patient_combined
variables_to_retrieve <- c(x1, x2, y1,y2)

# And store them as a tibble.
scores <- sample@meta.data[, variables_to_retrieve]
# Shuffle the cells so that we accomplish a random plotting, not sample by sample.

# Compute Y axis values.
d <- apply(scores, 1, function(x){max(x[c(x1, x2)]) - max(x[c(y1, y2)])})

# Compute X axis values.
x <- vapply(seq_along(d), function(x) {
  if (d[x] > 0) {
    d <- log2(abs(scores[x, x1] - scores[x, x2]) + 1)
    ifelse(scores[x, x1] < scores[x, x2], d, -d)
  } else {
    d <- log2(abs(scores[x, y1] - scores[x, y2]) + 1)
    ifelse(scores[x, y1] < scores[x, y2], d, -d)
  }
}, FUN.VALUE = numeric(1))

names(x) <- rownames(scores)

# Plot.
df_E <- data.frame(row.names = rownames(scores))
df_E[["set_x"]] <- x
df_E[["set_y"]] <- d
df_E[["group.by"]] <- sample@meta.data[, group.by]
df_E[["model"]] <- EP1NS_patient_combined$model


ggplot(df_E,aes(x=set_x,
              y=set_y,
              col=group.by))+
  geom_point(size=1.5)+
  scale_color_manual(values=col_model)+
   theme_cellstate_plot + 
  labs(x = 'Relative meta-module score', y = 'Relative meta-module score', title = 'EP1NS cellular hierarchy', subtitle = 'n = 919 cells') +
    geom_vline(xintercept = 0, linetype="dotted") + geom_hline(yintercept = 0, linetype="dotted") 
  #xlim(-2, 1) + ylim(-1, 3)

ggsave(file.path(plot_dir, "9_CellState_plot_4way_Metaprograms_patients.pdf"), width=5.5, height=4)

```